<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Apply ctSVG to Visium HD data â€¢ ctSVG</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Apply ctSVG to Visium HD data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ctSVG</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Visium%20HD.html">Apply ctSVG to Visium HD data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/haotian-zhuang/ctSVG/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Apply ctSVG to Visium HD data</h1>
                        <h4 data-toc-skip class="author">Haotian
Zhuang</h4>
            
            <h4 data-toc-skip class="date">October 27, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/haotian-zhuang/ctSVG/blob/HEAD/vignettes/Visium%20HD.Rmd" class="external-link"><code>vignettes/Visium HD.Rmd</code></a></small>
      <div class="d-none name"><code>Visium HD.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>We will demonstrate how to use ctSVG to generate single-cell gene
expression profiles and identify cell-type-specific SVGs from Visium HD
data. The Visium HD dataset of a mouse brain tissue was downloaded from
the <a href="https://www.10xgenomics.com/datasets/visium-hd-cytassist-gene-expression-libraries-of-mouse-brain-he" class="external-link">10x
website</a>. For analyzing Visium HD data, ctSVG requires two inputs: <a href="https://cf.10xgenomics.com/samples/spatial-exp/3.0.0/Visium_HD_Mouse_Brain/Visium_HD_Mouse_Brain_binned_outputs.tar.gz" class="external-link">10x
Space Ranger outputs</a>, and nuclei segmentation results.</p>
</div>
<div class="section level2">
<h2 id="nuclei-segmentation">Nuclei segmentation<a class="anchor" aria-label="anchor" href="#nuclei-segmentation"></a>
</h2>
<p>The nuclei segmentation results can be obtained by running
segmentation methods, such as <a href="https://github.com/stardist/stardist" class="external-link"><strong>StarDist</strong></a>,
on the H&amp;E images accompanying the Visium HD data. Here, the
microscope image of the mouse brain tissue was downloaded from the <a href="https://cf.10xgenomics.com/samples/spatial-exp/3.0.0/Visium_HD_Mouse_Brain/Visium_HD_Mouse_Brain_tissue_image.tif" class="external-link">10x
website</a>.</p>
<pre><code>import pandas as pd
import numpy as np
from tifffile import imread, imwrite
from csbdeep.utils import normalize
from stardist.models import StarDist2D

img = imread('Visium_HD_Mouse_Brain_tissue_image.tif')

# Load the pretrained model
model = StarDist2D.from_pretrained('2D_versatile_he')

# Percentile normalization of the image
min_percentile = 5
max_percentile = 95
img = normalize(img, min_percentile, max_percentile)

# Stardist
labels, polys = model.predict_instances_big(img, axes='YXC', block_size=4096, prob_thresh=0.01,nms_thresh=0.001, min_overlap=128, context=128, normalizer=None, n_tiles=(4,4,1))

# Convert data format
data = [(cell_id, x, y) for cell_id, (xs, ys) in enumerate(polys['coord']) for x, y in zip(xs, ys)]
df = pd.DataFrame(data, columns=['Cell', 'X', 'Y'])

df.to_csv('nuclei.csv', index=False)</code></pre>
</div>
<div class="section level2">
<h2 id="load-packages-and-datasets">Load packages and datasets<a class="anchor" aria-label="anchor" href="#load-packages-and-datasets"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://haotian-zhuang.github.io/ctSVG/">ctSVG</a></span><span class="op">)</span></span>
<span><span class="va">mydir</span> <span class="op">&lt;-</span> <span class="st">"~/Desktop/Research/ctSVG/mouse_brain"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">mydir</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "binned_outputs" "nuclei.csv"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="assign-each-square-to-a-cell">Assign each square to a cell<a class="anchor" aria-label="anchor" href="#assign-each-square-to-a-cell"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>  <span class="va">cellassignRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cellassign.html">cellassign</a></span><span class="op">(</span>visiumpath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">mydir</span>, <span class="st">"binned_outputs/square_002um"</span><span class="op">)</span>,</span>
<span>                              segmentationpath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">mydir</span>, <span class="st">"nuclei.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">cellassignRes</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ assign: Named chr [1:1858547] "Cell:21001" "Cell:21001" "Cell:21001" "Cell:21001" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:1858547] "s_002um_00029_00776-1" "s_002um_00029_00777-1" "s_002um_00029_00778-1" "s_002um_00029_00779-1" ...</span></span>
<span><span class="co">#&gt;  $ coord : num [1:82357, 1:2] 355 365 363 364 362 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:82357] "Cell:21001" "Cell:29526" "Cell:19243" "Cell:31821" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:2] "row" "col"</span></span>
<span><span class="va">coord</span> <span class="op">&lt;-</span> <span class="va">cellassignRes</span><span class="op">$</span><span class="va">coord</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="generate-single-cell-gene-expression-profiles">Generate single-cell gene expression profiles<a class="anchor" aria-label="anchor" href="#generate-single-cell-gene-expression-profiles"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getexpr.html">getexpr</a></span><span class="op">(</span>cellassignRes <span class="op">=</span> <span class="va">cellassignRes</span>,</span>
<span>                 visiumpath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">mydir</span>, <span class="st">"binned_outputs/square_002um"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span></span>
<span><span class="co">#&gt; Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;   ..@ i       : int [1:33493039] 42 109 198 243 291 364 561 627 744 763 ...</span></span>
<span><span class="co">#&gt;   ..@ p       : int [1:82358] 0 233 265 373 416 436 476 510 623 694 ...</span></span>
<span><span class="co">#&gt;   ..@ Dim     : int [1:2] 19059 82357</span></span>
<span><span class="co">#&gt;   ..@ Dimnames:List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:19059] "Xkr4" "Rp1" "Sox17" "Lypla1" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:82357] "Cell:21001" "Cell:29526" "Cell:19243" "Cell:31821" ...</span></span>
<span><span class="co">#&gt;   ..@ x       : num [1:33493039] 1 1 2 1 5 1 1 1 3 1 ...</span></span>
<span><span class="co">#&gt;   ..@ factors : list()</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-seurat-pipeline">Run Seurat pipeline<a class="anchor" aria-label="anchor" href="#run-seurat-pipeline"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>  <span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runseurat.html">runseurat</a></span><span class="op">(</span>count <span class="op">=</span> <span class="va">count</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="reassign-cells-to-clusters">Reassign cells to clusters<a class="anchor" aria-label="anchor" href="#reassign-cells-to-clusters"></a>
</h2>
<p>In this vignette, <code>n.max</code> is set to 200 to reduce
computational time. By default, <code>n.max</code> is set to 1,000 in
practice.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">recluRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/recluster.html">recluster</a></span><span class="op">(</span>d <span class="op">=</span> <span class="va">d</span>, n.max <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">recluRes</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "jac.mat"  "clu.list"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="identify-spatially-variable-genes-within-each-cell-cluster">Identify spatially variable genes within each cell cluster<a class="anchor" aria-label="anchor" href="#identify-spatially-variable-genes-within-each-cell-cluster"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>  <span class="va">clutestRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ctsvg_test.html">ctsvg_test</a></span><span class="op">(</span>d <span class="op">=</span> <span class="va">d</span>, coord <span class="op">=</span> <span class="va">coord</span>,</span>
<span>                           clu.list <span class="op">=</span> <span class="va">recluRes</span><span class="op">$</span><span class="va">clu.list</span>,</span>
<span>                           jac.mat <span class="op">=</span> <span class="va">recluRes</span><span class="op">$</span><span class="va">jac.mat</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">clutestRes</span><span class="op">)</span></span>
<span><span class="co">#&gt;           cluster    gene           fdr          pval fdr.empirical</span></span>
<span><span class="co">#&gt; 0.C1ql3         0   C1ql3 1.557701e-293 1.535740e-297    0.05422556</span></span>
<span><span class="co">#&gt; 0.Snap25        0  Snap25 1.021168e-212 2.013542e-216    0.05422556</span></span>
<span><span class="co">#&gt; 0.Rab3c         0   Rab3c 1.371994e-184 4.057955e-188    0.05422556</span></span>
<span><span class="co">#&gt; 0.Nptxr         0   Nptxr 8.103704e-171 3.195782e-174    0.05422556</span></span>
<span><span class="co">#&gt; 0.Cyp26b1       0 Cyp26b1 1.441156e-165 7.104190e-169    0.05422556</span></span>
<span><span class="co">#&gt; 0.Man1a         0   Man1a 3.105247e-144 1.836881e-147    0.05422556</span></span>
<span><span class="co">#&gt;           pval.empirical fstat.ori</span></span>
<span><span class="co">#&gt; 0.C1ql3       0.00990099  78.98951</span></span>
<span><span class="co">#&gt; 0.Snap25      0.00990099  65.73079</span></span>
<span><span class="co">#&gt; 0.Rab3c       0.00990099  50.76282</span></span>
<span><span class="co">#&gt; 0.Nptxr       0.00990099  46.25440</span></span>
<span><span class="co">#&gt; 0.Cyp26b1     0.00990099  79.04013</span></span>
<span><span class="co">#&gt; 0.Man1a       0.00990099  41.00661</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fit-spatial-gene-expression-patterns-within-each-cell-cluster">Fit spatial gene expression patterns within each cell cluster<a class="anchor" aria-label="anchor" href="#fit-spatial-gene-expression-patterns-within-each-cell-cluster"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clufitRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ctsvg_fit.html">ctsvg_fit</a></span><span class="op">(</span>d <span class="op">=</span> <span class="va">d</span>, coord <span class="op">=</span> <span class="va">coord</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualize-spatial-patterns-of-a-single-gene">Visualize spatial patterns of a single gene<a class="anchor" aria-label="anchor" href="#visualize-spatial-patterns-of-a-single-gene"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGene.html">plotGene</a></span><span class="op">(</span>d <span class="op">=</span> <span class="va">d</span>, clufitRes <span class="op">=</span> <span class="va">clufitRes</span>, coord <span class="op">=</span> <span class="va">coord</span>, gene <span class="op">=</span> <span class="va">clutestRes</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>         cluster <span class="op">=</span> <span class="st">"0"</span>, background <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visium%20HD_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="identify-gene-modules-within-each-cell-cluster">Identify gene modules within each cell cluster<a class="anchor" aria-label="anchor" href="#identify-gene-modules-within-each-cell-cluster"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">moduleRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ctsvg_module.html">ctsvg_module</a></span><span class="op">(</span>d <span class="op">=</span> <span class="va">d</span>, clutestRes <span class="op">=</span> <span class="va">clutestRes</span>, clufitRes <span class="op">=</span> <span class="va">clufitRes</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualize-spatial-patterns-of-metagenes-within-a-cell-cluster">Visualize spatial patterns of metagenes within a cell cluster<a class="anchor" aria-label="anchor" href="#visualize-spatial-patterns-of-metagenes-within-a-cell-cluster"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMetagene.html">plotMetagene</a></span><span class="op">(</span>d <span class="op">=</span> <span class="va">d</span>, clufitRes <span class="op">=</span> <span class="va">clufitRes</span>, moduleRes <span class="op">=</span> <span class="va">moduleRes</span>, coord <span class="op">=</span> <span class="va">coord</span>,</span>
<span>             cluster <span class="op">=</span> <span class="st">"0"</span>, background <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visium%20HD_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS 15.0.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/New_York</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ctSVG_1.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3     rstudioapi_0.16.0      jsonlite_1.8.8        </span></span>
<span><span class="co">#&gt;   [4] magrittr_2.0.3         spatstat.utils_3.1-0   farver_2.1.2          </span></span>
<span><span class="co">#&gt;   [7] rmarkdown_2.28         zlibbioc_1.50.0        fs_1.6.4              </span></span>
<span><span class="co">#&gt;  [10] ragg_1.3.2             vctrs_0.6.5            ROCR_1.0-11           </span></span>
<span><span class="co">#&gt;  [13] spatstat.explore_3.3-2 S4Arrays_1.4.1         htmltools_0.5.8.1     </span></span>
<span><span class="co">#&gt;  [16] findPC_1.0             SparseArray_1.4.8      sass_0.4.9            </span></span>
<span><span class="co">#&gt;  [19] sctransform_0.4.1      parallelly_1.38.0      KernSmooth_2.23-24    </span></span>
<span><span class="co">#&gt;  [22] bslib_0.8.0            htmlwidgets_1.6.4      desc_1.4.3            </span></span>
<span><span class="co">#&gt;  [25] ica_1.0-3              plyr_1.8.9             plotly_4.10.4         </span></span>
<span><span class="co">#&gt;  [28] zoo_1.8-12             cachem_1.1.0           igraph_2.0.3          </span></span>
<span><span class="co">#&gt;  [31] mime_0.12              lifecycle_1.0.4        pkgconfig_2.0.3       </span></span>
<span><span class="co">#&gt;  [34] Matrix_1.7-0           R6_2.5.1               fastmap_1.2.0         </span></span>
<span><span class="co">#&gt;  [37] MatrixGenerics_1.16.0  fitdistrplus_1.2-1     future_1.34.0         </span></span>
<span><span class="co">#&gt;  [40] shiny_1.9.1            digest_0.6.37          colorspace_2.1-1      </span></span>
<span><span class="co">#&gt;  [43] S4Vectors_0.42.1       patchwork_1.3.0        Seurat_5.1.0          </span></span>
<span><span class="co">#&gt;  [46] tensor_1.5             RSpectra_0.16-2        irlba_2.3.5.1         </span></span>
<span><span class="co">#&gt;  [49] textshaping_0.4.0      labeling_0.4.3         progressr_0.14.0      </span></span>
<span><span class="co">#&gt;  [52] fansi_1.0.6            spatstat.sparse_3.1-0  httr_1.4.7            </span></span>
<span><span class="co">#&gt;  [55] polyclip_1.10-7        abind_1.4-8            compiler_4.4.1        </span></span>
<span><span class="co">#&gt;  [58] proxy_0.4-27           withr_3.0.1            bit64_4.5.2           </span></span>
<span><span class="co">#&gt;  [61] DBI_1.2.3              fastDummies_1.7.4      highr_0.11            </span></span>
<span><span class="co">#&gt;  [64] MASS_7.3-60.2          rjson_0.2.23           classInt_0.4-10       </span></span>
<span><span class="co">#&gt;  [67] units_0.8-5            tools_4.4.1            lmtest_0.9-40         </span></span>
<span><span class="co">#&gt;  [70] httpuv_1.6.15          future.apply_1.11.2    goftest_1.2-3         </span></span>
<span><span class="co">#&gt;  [73] glue_1.7.0             nlme_3.1-164           promises_1.3.0        </span></span>
<span><span class="co">#&gt;  [76] grid_4.4.1             sf_1.0-18              Rtsne_0.17            </span></span>
<span><span class="co">#&gt;  [79] cluster_2.1.6          reshape2_1.4.4         generics_0.1.3        </span></span>
<span><span class="co">#&gt;  [82] hdf5r_1.3.11           gtable_0.3.5           spatstat.data_3.1-2   </span></span>
<span><span class="co">#&gt;  [85] class_7.3-22           tidyr_1.3.1            data.table_1.16.2     </span></span>
<span><span class="co">#&gt;  [88] XVector_0.44.0         sp_2.1-4               utf8_1.2.4            </span></span>
<span><span class="co">#&gt;  [91] BiocGenerics_0.50.0    spatstat.geom_3.3-3    RcppAnnoy_0.0.22      </span></span>
<span><span class="co">#&gt;  [94] ggrepel_0.9.6          RANN_2.6.2             pillar_1.9.0          </span></span>
<span><span class="co">#&gt;  [97] stringr_1.5.1          spam_2.11-0            RcppHNSW_0.6.0        </span></span>
<span><span class="co">#&gt; [100] later_1.3.2            splines_4.4.1          dplyr_1.1.4           </span></span>
<span><span class="co">#&gt; [103] lattice_0.22-6         bit_4.5.0              survival_3.6-4        </span></span>
<span><span class="co">#&gt; [106] deldir_2.0-4           tidyselect_1.2.1       miniUI_0.1.1.1        </span></span>
<span><span class="co">#&gt; [109] pbapply_1.7-2          knitr_1.48             gridExtra_2.3         </span></span>
<span><span class="co">#&gt; [112] IRanges_2.38.1         scattermore_1.2        stats4_4.4.1          </span></span>
<span><span class="co">#&gt; [115] xfun_0.47              matrixStats_1.3.0      stringi_1.8.4         </span></span>
<span><span class="co">#&gt; [118] lazyeval_0.2.2         yaml_2.3.10            evaluate_0.24.0       </span></span>
<span><span class="co">#&gt; [121] codetools_0.2-20       tibble_3.2.1           cli_3.6.3             </span></span>
<span><span class="co">#&gt; [124] uwot_0.2.2             arrow_17.0.0.1         xtable_1.8-4          </span></span>
<span><span class="co">#&gt; [127] reticulate_1.39.0      systemfonts_1.1.0      munsell_0.5.1         </span></span>
<span><span class="co">#&gt; [130] jquerylib_0.1.4        Rcpp_1.0.13            globals_0.16.3        </span></span>
<span><span class="co">#&gt; [133] spatstat.random_3.3-2  png_0.1-8              spatstat.univar_3.0-1 </span></span>
<span><span class="co">#&gt; [136] parallel_4.4.1         assertthat_0.2.1       pkgdown_2.1.0         </span></span>
<span><span class="co">#&gt; [139] ggplot2_3.5.1          dotCall64_1.2          listenv_0.9.1         </span></span>
<span><span class="co">#&gt; [142] viridisLite_0.4.2      e1071_1.7-16           scales_1.3.0          </span></span>
<span><span class="co">#&gt; [145] ggridges_0.5.6         crayon_1.5.3           SeuratObject_5.0.2    </span></span>
<span><span class="co">#&gt; [148] leiden_0.4.3.1         purrr_1.0.2            rlang_1.1.4           </span></span>
<span><span class="co">#&gt; [151] cowplot_1.1.3</span></span></code></pre></div>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Haotian Zhuang, Zhicheng Ji.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
